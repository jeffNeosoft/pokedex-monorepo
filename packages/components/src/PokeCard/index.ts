export {default as PokeCard} from './PokeCard';
export * from "./PokeCard.types"