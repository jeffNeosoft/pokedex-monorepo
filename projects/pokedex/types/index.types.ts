import { TPokemonTypeKeys } from "@pokedex/components";
export interface IResult {
  name: string;
  id: number | string;
}
export interface IPageModel {
  page: number;
  pageSize: number;
}
export interface IPokemonAPIList {
  data: IResult[];
  count: number;
  cacheData: IResult[][];
  isLoading: boolean;
  pagination:IPageModel;
  error: string | null;
} ;
export interface IPokeList {
  count: number;
  results: IResult[];
}
export interface IFetchPokemonListProps {
  page: number;
  limit?: number;
}
export type IPokemonData = {
  id: number;
  name: string;
  height?: number;
  weight?: number;
  image?: string;
  stats: {
    name: string;
    value: number | string;
  }[];
  types?: TPokemonTypeKeys[];
};
