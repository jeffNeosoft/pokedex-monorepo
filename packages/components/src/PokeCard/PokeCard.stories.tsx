import React from "react";
import { Grid } from "@mui/material";
import PokeCard from "./PokeCard";
import { TPokemonTypeKeys } from "./PokeCard.types";
const cardProps = {
  name: "Pikachu",
  height: 10,
  weight: 20,
  image:
    "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/25.svg",
  stats: [
    {
      value: 10,
      name: "Power",
    },
    {
      value: 650,

      name: "HP",
    },
    {
      value: 650,

      name: "Power",
    },
    {
      value: 650,

      name: "HP",
    },
    {
      value: 650,

      name: "Power",
    },
  ],
  types: ["normal", "electric", "fire"] as TPokemonTypeKeys[],
};
export default {
  component: <PokeCard {...cardProps} />,
  title: "PokeCard",
};

export const Default = {
  render: () => (
    <div style={{ display: "flex" }}>
      <Grid
        container
        justifyContent={"center"}
        alignContent={"stretch"}
        alignItems={"stretch"}
      >
        <PokeCard {...cardProps} />
      </Grid>
    </div>
  )
};
