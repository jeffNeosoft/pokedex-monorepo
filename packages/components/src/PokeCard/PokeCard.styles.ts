import { TPokemonTypeKeys } from "./PokeCard.types";

export const POKEMON_TYPE = {
  bug: "#26de81",
  dragon: "#ffeaa7",
  electric: "#fed330",
  fairy: "#FF0069",
  fighting: "#30336b",
  fire: "#f0932b",
  flying: "#81ecec",
  grass: "#00b894",
  ground: "#EFB549",
  ghost: "#a55eea",
  ice: "#74b9ff",
  normal: "#95afc0",
  poison: "#6c5ce7",
  psychic: "#a29bfe",
  rock: "#2d3436",
  water: "#0190FF",
} as const;
export const styles = {
  muiCard: {
    height: "75vh",
    width: "70vw",
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center",
    padding: "2rem",
    outline: "1px solid black",
    outlineOffset: "-1rem",
    background: "rgba( 255, 255, 255, 0.35 )",
    boxShadow: "0 8px 32px 0 rgba( 31, 38, 135, 0.37 )",
    backdropFilter: "blur( 4.5px )",
    WebkitBackdropFilter: "blur( 4.5px )",
    borderRadius: "10px",
    border: "1px solid rgba( 255, 255, 255, 0.18 )",
  },
  mediaContainer: (types: TPokemonTypeKeys[]) => {
    let percent = 0;
    const lineraBackground = types.map((ele) => {
      const prev = percent;
      percent += 100 / types.length;
      return `${POKEMON_TYPE[ele]}80 ${prev}% ${percent}%`;
    });
    return {
      width: "100%",
      objectFit: "fill",
      height: "95%",
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      background: `linear-gradient(45deg,${lineraBackground.toString()})`,
      borderRadius: "10px",
    };
  },
  mediaCard:(laoding:boolean)=>{
    return {
      width: laoding ? "0px":"150px",
      objectFit: "fill",
      filter: "drop-shadow(5px 5px 5px rgba(0,0,0,0.8))",
    }
  } ,
  circularCard: {
    height: "65vh",
    margin: "0 auto",
    position: "absolute",
    top:"2rem",
    left:"calc(50% - 50px)",
    textAlign: "center",
    width: "200px",
    WebkitTransformOrigin: "center center",
    MozTransformOrigin: "center center",
    OTransformOrigin: "center center",
    MsTransformOrigin: "center center",
    transformOrigin: "center center",
  },
  cardContainer: (cardLength) => {
    const cardDegrees = 360 / cardLength;
    const cardCss = Array.from({ length: cardLength }, (c, i) => i).reduce(
      (acc, curr) => {
        return {
          ...acc,
          [`&:nth-child(${curr + 1})`]: {
            WebkitTransform: `rotate(${(curr + 1) * cardDegrees}deg)`,
            MozTransform: `rotate(${(curr + 1) * cardDegrees}deg)`,
            MsTransform: `rotate(${(curr + 1) * cardDegrees}deg)`,
            OTransform: `rotate(${(curr + 1) * cardDegrees}deg)`,
            transform: `rotate(${(curr + 1) * cardDegrees}deg)`,
            "& p":{
              rotate:`-${(curr + 1) * cardDegrees}deg`
            },
          },
        };
      },
      {}
    );

    return {
      background: "rgb(130,90,164)",
      border: "8px solid white",
      WebkitBorderRadius: "4px",
      MozBorderRadius: "4px",
      borderRadius: "4px",
      WebkitBoxShadow: "0 1px 3px rgba(0,0,0,.24)",
      MozBoxShadow: "0 1px 3px rgba(0,0,0,.24)",
      boxShadow: "0 1px 3px rgba(0,0,0,.24)",
      WebkitBoxSizing: "border-box",
      MozBoxSizing: "border-box",
      boxSizing: "border-box",
      height: "120px",
      top: "0",
      position: "absolute",
      width: "90px",
      WebkitTransformOrigin: "center 250px",
      MozTransformOrigin: "center 280px",
      OTransformOrigin: "center 280px",
      MsTransformOrigin: "center 280px",
      transformOrigin: "center 280px",
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      ...cardCss,
    };
  },
};
