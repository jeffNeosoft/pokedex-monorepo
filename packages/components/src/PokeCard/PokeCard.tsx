import React, { useState } from "react";
import MuiCard from "@mui/material/Card";
import { Box, CardMedia, Typography } from "@mui/material";
import { ICardProps } from "./PokeCard.types";
import BadgeComponent from "../BadgeComponent/BadgeComponent";
import { POKEMON_TYPE, styles } from "./PokeCard.styles";
import PokeLoader from "../PokeLoader/PokeLoader";

const Card = (props: ICardProps) => {
  const [loading, setLoading] = useState<boolean>(true);
  return (
    <MuiCard sx={styles.muiCard}>
      <Box sx={styles.mediaContainer(props.types)}>
        {loading && <PokeLoader />}
        <CardMedia
          sx={styles.mediaCard(loading)}
          component="img"
          image={props.image}
          title={props.name}
          loading="lazy"
          onLoad={() => setLoading(false)}
          onError={() => setLoading(false)}
        />
      </Box>
      <Box sx={styles.circularCard}>
        {props.stats &&
          props.stats.map((data,i) => (
            <Box key={`statCard${i}`} sx={styles.cardContainer(props.stats.length)}>
              <Typography variant={"caption"} color={"white"}>
                <p>{data.name}</p>
                <p>{data.value}</p>
              </Typography>
            </Box>
          ))}
      </Box>
      <Typography
        gutterBottom
        variant="h4"
        component="div"
        sx={{
          textTransform: "capitalize",
          textAlign: "center",
        }}
      >
        {props.name}
      </Typography>
      <Box>
        {props.types.map((type,i) => (
          <Box key={`type${i}`} sx={{ display: "inline-block" , margin:"1rem"}}>
            <BadgeComponent
              labelBackground={POKEMON_TYPE[type]}
              label={type}
              labelFontColor="white"
            />
          </Box>
        ))}
      </Box>
    </MuiCard>
  );
};
export default Card;
