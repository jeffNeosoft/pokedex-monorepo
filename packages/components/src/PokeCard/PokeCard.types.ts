import { POKEMON_TYPE } from "./PokeCard.styles";

export type TPokemonTypeKeys = keyof typeof POKEMON_TYPE;
export type IAbilityProp = {
	ability: {
		name: string;
		url: string;
	};
};
export interface ICardProps {
	name: string;
	height?: number;
	weight?: number;
	image?: string;
	stats?: {
		value: number | string;
		name: string
	}[];
	types?: TPokemonTypeKeys[];
}

