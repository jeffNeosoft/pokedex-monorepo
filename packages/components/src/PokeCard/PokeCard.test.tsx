import * as React from "react";
import { render, screen } from "@testing-library/react";
import Card from "./PokeCard";
import { TPokemonTypeKeys } from "./PokeCard.types";

describe("Card component", () => {
  const mockProps = {
    name: "Raichu",
    image: "Raichu.jpg",
    height: 23,
    weight: 34,
    location: "https://pokeapi.co/api/v2/pokemon/25",
    stats: [
      {
        name: "attack",
        value: 10,
      },
    ],
    types: ["poison"] as TPokemonTypeKeys[],
  };

  it("Renders the card successfully", () => {
    const { container } = render(<Card {...mockProps} />);

    expect(screen.getByText("Raichu")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
