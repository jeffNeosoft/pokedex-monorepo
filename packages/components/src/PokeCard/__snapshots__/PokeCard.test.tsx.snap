// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component Renders the card successfully 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1ms03a4-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiBox-root css-1q75yu9"
    >
      <div
        class="MuiBox-root css-k9rf7z"
      >
        <div
          class="MuiBox-root css-z3zg3m"
        >
          <div
            class="MuiBox-root css-syuk68"
            data-testid="pokeball"
          >
            <div
              class="MuiBox-root css-wlkd4e"
            />
          </div>
        </div>
      </div>
      <img
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1x1fx0l-MuiCardMedia-root"
        loading="lazy"
        src="Raichu.jpg"
        title="Raichu"
      />
    </div>
    <div
      class="MuiBox-root css-14pk3wm"
    >
      <div
        class="MuiBox-root css-7175pe"
      >
        <span
          class="MuiTypography-root MuiTypography-caption css-190nz9y-MuiTypography-root"
        >
          <p>
            attack
          </p>
          <p>
            10
          </p>
        </span>
      </div>
    </div>
    <div
      class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-qj0d3u-MuiTypography-root"
    >
      Raichu
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-1ggw9ik"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
        >
          <div
            class="MuiBox-root css-1cxz5bq"
          >
            poison
          </div>
        </p>
      </div>
    </div>
  </div>
</div>
`;
